src/_zkapauthorizer/resource.py

```diff
@@ -214,110 +213,174 @@ def _setup_replication(self, request):
         request.finish()
 
 
-@define
-class RecoverResource(Resource):
+class RecoverProtocol(WebSocketServerProtocol):
     """
-    Implement the endpoint for triggering local state recovery from a remote
-    replica.
+    Speaks the server side of the WebSocket /recover protocol.
+
+    A client connects to this to start recovery, sending an opening
+    message with the rquired capability.
+
+    As recovery is ongoing, the server sends status updates as they
+    become available.
+
+    When the recovery is finished, and final message is sent
+    (indicating overall success or failure) and the WebSocket is
+    closed.
     """
 
-    _log = Logger()
+    def onClose(self, wasClean, code, reason) -> None:
+        """
+        WebSocket API: we've lost our connection for some reason
+        """
+        try:
+            self.factory.clients.remove(self)
+        except ValueError:
+            # may not have initiated recovery yet so it might not be
+            # in the clients list
+            pass
+
+    def onMessage(self, payload, isBinary) -> None:
+        """
+        WebSocket API: a message has been received from the client (the
+        only thing they can send is a request to initiate recovery).
+        """
+        try:
+            body = loads(payload)
+            if set(body.keys()) != {"recovery-capability"}:
+                raise ValueError("Unknown keys present in request")
+            recovery_capability = from_string(body["recovery-capability"])
+            if not isinstance(recovery_capability, ReadonlyDirectoryURI):
+                raise ValueError("Not a readonly-dircap")
+        except Exception:
+            self.sendClose(
+                code=4000,
+                reason="Failed to parse recovery request",
+            )
+            return
+        # we have a valid request, tell our factory to start recovery
+        self.factory.initiate_recovery(recovery_capability, self)
+
+
+@define
+class RecoverFactory(WebSocketServerFactory):
+    """
+    Track state of recovery.
+
+    In the factory because we want at most one active recovery attempt
+    no matter how many clients there are and because something needs
+    to link to other resources that are also constructed once.
+    """
 
     store: VoucherStore = field()
     get_downloader: Callable[[str], Downloader] = field()
-    recoverer: StatefulRecoverer = field(default=Factory(StatefulRecoverer))
+    recoverer: StatefulRecoverer = field()
+    recovering_d: Deferred = field(default=None)
+    recovering_cap: CapStr = field(default=None)
```

Comment:
in `initiate_recovery` `recovering_cap` is set to a `ReadonlyDirectoryURI`... FWIW it looks like `ReadonlyDirectoryURI` is the type actually used at runtime.

Also, these two recovering fields should actually be `Optional[...]`?  I dunno wtf is wrong with mypy.

Does subclassing something from autobahn just break all mypy inference for this whole class?